### PR TITLE
feat(dashboards): Infinite duplicate dashboard names

### LINKED ISSUE #77430: Add Dashboard button in templates does not work in some cases
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

1. Go to Dashboards and click on the Add dashboards button for any one of the 4 templates
2. Repeat this 9 times (until 9 copies of the same template dashboard is created)
3. Repeat it one last time (10th copy)

Image

### Expected Result

A new copy of the template is created and named something like "General Template copy 10"

### Actual Result

Error saying 'Dashboard title already taken'

Image

### Product Area

Dashboards

### Link

_No response_

### DSN

_No response_

### Version

_No response_

### PR DESCRIPTION
**Pull request by gggritso**
## tl;dr

Right now if you duplicate too many dashboards, or make too many of a dashboard template, you get an error. People don't like that. This PR makes it so you can make infinite duplicates, if you want.

## Changes

We've had a few complaints (and a few Sentry exceptions) from people who are clicking the "Duplicate" button on a dashboard and/or clicking the "Add Dashboard" button on the Dashboards page and seeing errors.

When people add copies of a dashboard we automatically increment the name. The sequence goes "My Dashboard", "My Dashboard copy", "My Dashboard copy 1", "My Dashboard copy 2", etc. We cap out at 9. Beyond 9 we show a little error toast. This is fine but isn't _awesome_. Also, a quirk of the endpoint code is that it recursively calls `self.post` and tries to insert a dashboard. For high copy numbers, this means 10 failed inserts in a row, which is not the most efficient.

This PR changes the approach. When a dashboard is created, it attempts an insert. If the insert fails due to `IntegrityError`, it calls `Dashboard.incremental_title` and fetches a "guaranteed" unique incremented title, and tries again.

The "guaranteed" unique title is done by fetching all matching dashboard names from the database, and finding the highest increment that way. This is subtly different from the current behaviour, because if someone deletes "My Dashboard copy 1" and duplicates "My Dashboard", they'll get "My Dashboard copy 3" if "My Dashboard copy 2" still exists. IMO this is better because that's closer to how many copies have been made.

## Performance

I checked the stats and the p99 for number of dashboards for an org is 16. The p100 is ~400 which is high but still pretty fast to find the right number. In the worst case this takes about 5ms. This scan only happens in case of a conflict, anyway, which is rare.

Closes https://github.com/getsentry/sentry/issues/77430

### CODE REVIEW

**Review by nikkikapadia — APPROVED** [review]

File: src/sentry/models/dashboard.py [hunk 1]
```diff
@@ -77,6 +78,32 @@ def get_prebuilt(organization, user, dashboard_id):
...
+    @classmethod
+    def incremental_title(cls, organization, name):
+        """
+        Given a dashboard name that migh already exist, returns a new unique name that does not exist, by appending the word "copy" and an integer if necessary.
+        """
+
+        base_name = re.sub(r" ?copy ?(\d+)?$", "", name)
+        matching_dashboards = cls.objects.filter(
+            organization=organization, title__regex=rf"^{re.escape(base_name)} ?(copy)? ?(\d+)?$"
+        )
```
Comment (narsaynorath) [inline src/sentry/models/dashboard.py:90]: Should the regex be _not_ optionally capturing the spaces, copy, and number? Since we know the intended format of the name we want to set, we can just query for ones that would potentially conflict (i.e. don't have to sift through the ones that won't conflict because copy and the number are optional)

```suggestion
        matching_dashboards = cls.objects.filter(
            organization=organization, title__regex=rf"^{re.escape(base_name)} copy \d+$"
        )
```

Comment (narsaynorath) [reply src/sentry/models/dashboard.py:90]: Also might be a tiny optimization, but you can slap a `.values_list("title")` in there to just get back an iterable of titles, but this is a low priority suggestion considering the number of dashboards there are

File: tests/sentry/models/test_dashboard.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from sentry.models.dashboard import Dashboard
+from sentry.testutils.cases import TestCase
+
+
+class IncrementalNameTest(TestCase):
```
Comment (narsaynorath) [inline tests/sentry/models/test_dashboard.py:5]: Since the original issue was to be able to successfully do a POST request with >10 dashboards, could we mock something similar in the endpoint tests to actually verify that? e.g. these tests have great coverage but it doesn't guarantee someone won't mess up how this is being used in the endpoint

**Review by narsaynorath — APPROVED** [review]
This looks good to me, I just had a comment around the regex we're using to get the "eligible" dashboards to loop through and adding a test to ensure the endpoint behaves like we intend

Great tests 🙏

File: src/sentry/models/dashboard.py (hunk 1 above)
Comment (gggritso) [reply src/sentry/models/dashboard.py]: 🤔 with a desired name "My Widgets", names "My Widgets", "My Widgets copy", and "My Widgets copy 1" are all conflicts, so I think I have to query for all of those, no? That's why all the spaces, "copy" and integer are all optional. Is there a way to condense that?

Added the `.values`, thanks for that! I'd rather have the dict than the tuple, I think

File: tests/sentry/models/test_dashboard.py (hunk 2 above)
Comment (gggritso) [reply tests/sentry/models/test_dashboard.py:5]: Added some more endpoint tests to cover this!

File: src/sentry/models/dashboard.py (hunk 1 above)
Comment (narsaynorath) [reply src/sentry/models/dashboard.py]: Ah good point, I was trying to avoid cases like "My Widgets 1" but it's such a small set of cases to encounter. 

I also forgot to mention but you can add `flat_list=True` to `values_list` to get rid of the tuples

**Review by narsaynorath — APPROVED** [review]
